### PR TITLE
GH-111485: Factor out generation of uop IDs from cases generator.

### PR DESCRIPTION
**Pull request by markshannon**
* Issue: gh-111485

### CODE REVIEW

File: Tools/cases_generator/uop_id_generator.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+def generate_uop_ids(filenames: str, analysis: Analysis, outfile: TextIO, distinct_namespace: bool) -> None:
+    write_header(__file__, filenames, outfile)
+    out = CWriter(outfile, 0, False)
+    out.emit(
+        """#ifndef Py_CORE_UOP_IDS_H
+#define Py_CORE_UOP_IDS_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+""")
```
Comment (gvanrossum) [inline Tools/cases_generator/uop_id_generator.py]: Add a blank line here.
```suggestion
#endif

""")
```

File: Tools/cases_generator/uop_id_generator.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    write_header,
+)
+from cwriter import CWriter
+from typing import TextIO
+
+
+DEFAULT_OUTPUT = ROOT / "Include/internal/pycore_uop_ids.h"
+
+
+def generate_uop_ids(filenames: str, analysis: Analysis, outfile: TextIO, distinct_namespace: bool) -> None:
```
Comment (gvanrossum) [inline Tools/cases_generator/uop_id_generator.py]: Run in through black?

File: Tools/cases_generator/uop_id_generator.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+arg_parser.add_argument(
+    "input", nargs=argparse.REMAINDER, help="Instruction definition file(s)"
+)
+
+if __name__ == "__main__":
+    args = arg_parser.parse_args()
+    if len(args.input) == 0:
+        args.input.append(DEFAULT_INPUT)
```
Comment (gvanrossum) [inline Tools/cases_generator/uop_id_generator.py]: It seems this adds a `Path` instance and something in formatting.py doesn't like that yet:
```
Traceback (most recent call last):
  File "/Users/guido/cpython/Tools/cases_generator/uop_id_generator.py", line 89, in <module>
    data = analyze_files(args.input)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/guido/cpython/Tools/cases_generator/analyzer.py", line 433, in analyze_files
    return analyze_forest(parser.parse_files(filenames))
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/guido/cpython/Tools/cases_generator/parser.py", line 26, in parse_files
    psr = Parser(src, filename=prettify_filename(filename))
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/guido/cpython/Tools/cases_generator/formatting.py", line 150, in prettify_filename
    filename = filename.replace("\\", "/")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: Path.replace() takes 2 positional arguments but 3 were given
```

Comment (gvanrossum) [reply Tools/cases_generator/uop_id_generator.py]: (Same for opcode_id_generator.py.)

File: Tools/cases_generator/generate_cases.py [hunk 4]
```diff
@@ -381,7 +381,7 @@ def write_metadata(self, metadata_filename: str, pymetadata_filename: str) -> No
             self.write_pseudo_instrs()
 
             self.out.emit("")
-            self.write_uop_items(lambda name, counter: f"#define {name} {counter}")
+            self.out.emit('#include "pycore_uop_ids.h"')
```
Comment (gvanrossum) [inline Tools/cases_generator/generate_cases.py:384]: This creates dead code (`write_uop_items` is never called). I suppose at the end you're planning a grand removal of all of generate_cases.py?

Comment (markshannon) [reply Tools/cases_generator/generate_cases.py:384]: Yes

File: Tools/cases_generator/uop_id_generator.py (hunk 3 above)
Comment (markshannon) [reply Tools/cases_generator/uop_id_generator.py]: Maybe we should drop the idea of a default input (or output)?
We need to make them explicit in the Makefile, so I see little advantage to having defaults.

File: Include/internal/pycore_uop_ids.h [hunk 5]
```diff
@@ -0,0 +1,268 @@
+// This file is generated by Tools/cases_generator/uop_id_generator.py
```
Comment (iritkatriel) [inline Include/internal/pycore_uop_ids.h:1]: This should be marked as generated in .gitattributes

File: Tools/cases_generator/uop_id_generator.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            out.emit(f"#define {uop.name} {uop.name[1:]}\n")
+        else:
+            out.emit(f"#define {uop.name} {next_id}\n")
+            next_id += 1
+
+    out.emit("\n")
+    out.emit("#ifdef __cplusplus\n")
+    out.emit("}\n")
+    out.emit("#endif\n")
+    out.emit("#endif /* !Py_OPCODE_IDS_H */\n")
```
Comment (iritkatriel) [inline Tools/cases_generator/uop_id_generator.py:63]: This postscript repeats itself in other generated files. Could be generated by a common utility function eventually.  Or even a common context manager that adds the opening and closing bits.

**Review by iritkatriel — APPROVED** [review]

File: Tools/cases_generator/uop_id_generator.py (hunk 6 above)
Comment (markshannon) [reply Tools/cases_generator/uop_id_generator.py:63]: Yes, that makes sense for the header files.

File: Tools/cases_generator/uop_id_generator.py (hunk 3 above)
Comment (gvanrossum) [reply Tools/cases_generator/uop_id_generator.py]: When developing this script, the defaults are very helpful.

**Pull request merged**